README.md
# bibaugust
Download: ![download badge](https://github.com/raphael-lesmana/bibaugust/actions/workflows/c-cpp.yml/badge.svg)


history.txt
2019-10-12#C in a Nutshell: The Definitive Reference#Aldo#P
2019-10-15#Learning Java: An Introduction to Real-World Programming with Java#Bima#P
2021-09-09#Test#Ludwig#K


load.c
#include "main.h"

int hitungBaris(char *filename)
{
    FILE *dbFile = fopen(filename, "r");
    int counter = 0;
    char tmp;

    while ((tmp = fgetc(dbFile)) != EOF)
    {
        if (tmp == '\n')
        {
            counter++;
        }
    }

    fclose(dbFile);
    return counter;
}

int binarySearch(Buku *daftarBuku, char *keyword, int min, int max)
{
    if (max >= min)
    {
        int mid = (min + max) / 2;
        if (strcmp(daftarBuku[mid].judul, keyword) == 0)
        {
            return mid;
        }
        if (strcmp(daftarBuku[mid].judul, keyword) < 0)
        {
            return binarySearch(daftarBuku, keyword, mid + 1, max);
        }
        return binarySearch(daftarBuku, keyword, min, mid - 1);
    }
    return -1;
}

void loadDatabase(char *filename, Buku *daftarBuku, int n)
{
    FILE *dbFile = fopen(filename, "r");

    for (int i = 0; i < n; i++)
    {
        fscanf(dbFile, "%[^#]#%i#%[^#]#%i#%lf#%lld#%f#%i\n", daftarBuku[i].judul,
                &daftarBuku[i].tahunTerbit, daftarBuku[i].penulis,
                &daftarBuku[i].halaman, &daftarBuku[i].berat,
                &daftarBuku[i].isbn, &daftarBuku[i].rating,
                &daftarBuku[i].dipinjam);
    }

    fclose(dbFile);
}

void loadRiwayat(char *filename, Riwayat *daftarRiwayat, int n)
{
    FILE *histFile = fopen(filename, "r");

    for (int i = 0; i < n; i++)
    {
        fscanf(histFile, "%[^#]#%[^#]#%[^#]#%c\n", daftarRiwayat[i].tanggal,
                daftarRiwayat[i].judul, daftarRiwayat[i].peminjam,
                &daftarRiwayat[i].keterangan);
    }

    fclose(histFile);
}

void updateDatabase(char *filename, Buku **daftarBuku, int *jumlahBuku)
{
    int jumlahTmp = hitungBaris(filename);
    *jumlahBuku = jumlahTmp;

    Buku *daftarTmp = realloc(*daftarBuku, jumlahTmp * sizeof(Buku));
    Buku tmp;

    if (daftarTmp == NULL)
    {
        printf("Terjadi kesalahan!");
        return;
    }

    FILE *dbFile = fopen(filename, "r");

    for (int i = 0; i < jumlahTmp; i++)
    {
        fscanf(dbFile, "%[^#]#%i#%[^#]#%i#%lf#%lld#%f#%i\n", tmp.judul,
                &tmp.tahunTerbit, tmp.penulis,
                &tmp.halaman, &tmp.berat,
                &tmp.isbn, &tmp.rating, &tmp.dipinjam);
    }

    fclose(dbFile);

    daftarTmp[jumlahTmp - 1] = tmp;
    *daftarBuku = daftarTmp;
}

void updateRiwayat(char *filename, Riwayat **daftarRiwayat, int *jumlahRiwayat)
{
    int jumlahTmp = hitungBaris(filename);
    *jumlahRiwayat = jumlahTmp;

    Riwayat *daftarTmp = realloc(*daftarRiwayat, jumlahTmp * sizeof(Riwayat));
    Riwayat tmp;

    if (daftarTmp == NULL)
    {
        printf("Terjadi kesalahan!");
        return;
    }

    FILE *histFile = fopen(filename, "r");

    for (int i = 0; i < jumlahTmp; i++)
    {
        fscanf(histFile, "%[^#]#%[^#]#%[^#]#%c\n", tmp.tanggal,
                tmp.judul, tmp.peminjam, &tmp.keterangan);
    }

    fclose(histFile);

    daftarTmp[jumlahTmp - 1] = tmp;
    *daftarRiwayat = daftarTmp;
}

main.c
#include "main.h"

int main(void)
{
    int jumlahBuku = hitungBaris("db.txt");
    Buku *daftarBuku = malloc(jumlahBuku * sizeof(Buku));
    Buku *daftarBukuUnsorted = malloc(jumlahBuku * sizeof(Buku));

    if (daftarBuku == NULL)
    {
        fprintf(stderr, "Telah terjadi kesalahan!");
    }

    loadDatabase("db.txt", daftarBuku, jumlahBuku);
    loadDatabase("db.txt", daftarBukuUnsorted, jumlahBuku);
    sort(daftarBuku, jumlahBuku);

    int jumlahRiwayat = hitungBaris("history.txt");
    Riwayat *daftarRiwayat = malloc(jumlahRiwayat * sizeof(Riwayat));
    if (daftarRiwayat == NULL)
    {
        fprintf(stderr, "Telah terjadi kesalahan!");
    }

    loadRiwayat("history.txt", daftarRiwayat, jumlahRiwayat);

    int input;
    char last_message[512];
    do
    {
        system("clear");
        printf("-----------------------------------\n");
        printf("|            BIBAUGUST            |\n");
        printf("|                                 |\n");
        printf("| 1. Cari buku                    |\n");
        printf("| 2. Pinjam buku                  |\n");
        printf("| 3. Kembalikan buku              |\n");
        printf("| 4. Tambah buku                  |\n");
        printf("| 5. Hapus buku                   |\n");
        printf("| 6. Riwayat peminjaman           |\n");
        printf("| 7. Daftar buku                  |\n");
        printf("| 8. Exit                         |\n");
        printf("-----------------------------------\n");
        printf("Masukkan angka dari menu: ");
        scanf("%i", &input);
        getchar();

        switch(input)
        {
            case 1:
                system("clear");
                cariBuku(daftarBuku, jumlahBuku);
                break;
            case 2:
                system("clear");
                pinjamBuku(&daftarBuku, &jumlahBuku, &daftarRiwayat, &jumlahRiwayat);
                break;
            case 3:
                system("clear");
                kembalikanBuku(&daftarBuku, &jumlahBuku, &daftarRiwayat, &jumlahRiwayat);
                break;
            case 4:
                system("clear");
                tambahBuku(&daftarBuku, &jumlahBuku);
                break;
            case 5:
                system("clear");
                hapusBuku(&daftarBukuUnsorted, &jumlahBuku);
                break;
            case 6:
                system("clear");
                lihatRiwayat(daftarRiwayat, jumlahRiwayat);
                break;
            case 7:
                system("clear");
                lihatDatabase(daftarBuku, jumlahBuku);
                break;
            case 8:
                break;
            default:
                printf("Maaf, angka %i bukan pilihan valid\n", input);
        }
    }
    while (input != 8);

    free(daftarBuku);
    free(daftarBukuUnsorted);
    free(daftarRiwayat);

    return 0;
}


main.h
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

typedef struct Buku
{
    char judul[128];
    int tahunTerbit;
    char penulis[64];
    int halaman;
    double berat;
    long long int isbn;
    float rating;
    bool dipinjam;
}
Buku;

typedef struct Riwayat
{
    char tanggal[32];
    char judul[128];
    char peminjam[64];
    char keterangan;
}
Riwayat;

// tools.c
void hapusBuku(Buku **daftarBuku, int *jumlahBuku);
void tambahBuku(Buku **daftarBuku, int *jumlahBuku);
void lihatRiwayat(Riwayat *daftarRiwayat, int jumlahRiwayat);
void cariBuku(Buku *daftarBuku, int jumlahBuku);
void lihatDatabase(Buku *daftarBuku, int jumlahBuku);
void kembalikanBuku(Buku **daftarBuku, int *jumlahBuku, Riwayat **daftarRiwayat, int *jumlahRiwayat);
void pinjamBuku(Buku **daftarBuku, int *jumlahBuku, Riwayat **daftarRiwayat, int *jumlahRiwayat);

// load.c
int hitungBaris(char *filename);
int binarySearch(Buku *daftarBuku, char *keyword, int min, int max);
void loadDatabase(char *filename, Buku daftarBuku[], int n);
void loadRiwayat(char *filename, Riwayat daftarRiwayat[], int n);
void updateDatabase(char *filename, Buku **daftarBuku, int *jumlahBuku);
void updateRiwayat(char *filename, Riwayat **daftarRiwayat, int *jumlahRiwayat);

// sort.c
void merge(Buku daftarBuku[], int left, int mid, int right);
void split(Buku daftarBuku[], int left, int right);
void sort(Buku daftarBuku[], int n);

tools.c
#include "main.h"

void hapusBuku(Buku **daftarBuku, int *jumlahBuku)
{
    int indexHapus = 0;
    char judul[128];
    printf("Masukkan nama buku yang ingin dihapus: ");
    scanf("%[^\n]", judul);
    getchar();

    if ((indexHapus = binarySearch(*daftarBuku, judul, 0, *jumlahBuku - 1)) == -1)
    {
        printf("Buku tidak ditemukan!\n");
        printf("Tekan tombol apa saja untuk kembali ke menu\n");
        getchar();
    }
    FILE *dbFile = fopen("db.txt", "r");
    FILE *tmpFile = fopen("tmp", "w");
    int i = 0;
    char buffer[512];

    while (fscanf(dbFile, "%[^\n]\n", buffer) != EOF)
    {
        if (i == indexHapus)
        {
            i++;
            continue;
        }
        fprintf(tmpFile, "%s\n", buffer);
        i++;
    }

    fclose(dbFile);
    fclose(tmpFile);

    dbFile = fopen("db.txt", "w");
    tmpFile = fopen("tmp", "r");

    while (fscanf(tmpFile, "%[^\n]\n", buffer) != EOF)
    {
        fprintf(dbFile, "%s\n", buffer);
    }

    fclose(dbFile);
    fclose(tmpFile);
    remove("tmp");

    updateDatabase("db.txt", daftarBuku, jumlahBuku);
    printf("Buku berhasil dihapus\n");
    printf("Tekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}

void tambahBuku(Buku **daftarBuku, int *jumlahBuku)
{
    Buku bukuTambahan;

    printf("Masukkan data buku yang ingin Anda tambahkan ke dalam database!\n");
    printf("Judul Buku: ");
    scanf("%[^\n]", bukuTambahan.judul);
    getchar();
    printf("Tahun Terbit: ");
    scanf("%d", &bukuTambahan.tahunTerbit);
    getchar();
    printf("Nama Penulis: ");
    scanf("%[^\n]", bukuTambahan.penulis);
    getchar();
    printf("Jumlah Halaman: ");
    scanf("%d", &bukuTambahan.halaman);
    getchar();
    printf("Berat Buku: ");
    scanf("%lf", &bukuTambahan.berat);
    getchar();
    printf("ISBN: ");
    scanf("%lld", &bukuTambahan.isbn);
    getchar();
    printf("Rating Buku: ");
    scanf("%f", &bukuTambahan.rating);
    getchar();
    printf("Dipinjam (Masukkan 1 = True, 0 = False): ");
    scanf("%i", &bukuTambahan.dipinjam);
    getchar();

    // menambahkan input user ke dalam db.txt
    FILE *dbFile = fopen("db.txt", "a+");
    fprintf(dbFile, "%s#%d#%s#%d#%lf#%lld#%f#%i\n", bukuTambahan.judul,
            bukuTambahan.tahunTerbit, bukuTambahan.penulis,
            bukuTambahan.halaman, bukuTambahan.berat, bukuTambahan.isbn,
            bukuTambahan.rating, bukuTambahan.dipinjam);

    fclose(dbFile);
    // update array
    updateDatabase("db.txt", daftarBuku, jumlahBuku);

    printf("Buku Anda telah ditambahkan ke database kami!\n");
    printf("Tekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}

void lihatRiwayat(Riwayat *daftarRiwayat, int jumlahRiwayat)
{
    for (int i = 0; i < jumlahRiwayat; i++)
    {

        printf("Tanggal: %s\n", daftarRiwayat[i].tanggal);
        printf("Judul: %s\n", daftarRiwayat[i].judul);
        printf("Peminjam: %s\n", daftarRiwayat[i].peminjam);
        printf("Keterangan: %s\n\n", (daftarRiwayat[i].keterangan == 'P') ? "Meminjam" : "Mengembalikan" );
    }
    printf("\nTekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}

void cariBuku(Buku *daftarBuku, int jumlahBuku)
{
    char input[128];
    printf("Masukkan judul buku yang ingin dicari: ");
    scanf("%[^\n]", input );
    getchar();
    int n;
    n = binarySearch(daftarBuku, input, 0, jumlahBuku - 1);
    if (n != -1)
    {
        printf("Judul: %s\n", daftarBuku[n].judul);
    	printf("Tahun Terbit: %d\n", daftarBuku[n].tahunTerbit);
    	printf("Penulis: %s\n", daftarBuku[n].penulis);
    	printf("Jumlah Halaman: %d\n", daftarBuku[n].halaman);
    	printf("Berat Buku: %lf\n", daftarBuku[n].berat);
    	printf("ISBN: %lld\n", daftarBuku[n].isbn);
    	printf("Rating: %f\n", daftarBuku[n].rating);
    	printf("Dipinjam? %s\n\n", (daftarBuku[n].dipinjam) ? "Iya" : "Tidak");
        printf("Tekan tombol apa saja untuk kembali ke menu\n");
        getchar();
    }
    else
    {
        printf("Buku tidak ditemukan!\n");
        printf("Tekan tombol apa saja untuk kembali ke menu\n");
        getchar();
    }
}

void lihatDatabase(Buku *daftarBuku, int jumlahBuku)
{
    for(int i=0; i<jumlahBuku; i++)
    {
    	printf("Judul: %s\n", daftarBuku[i].judul);
    	printf("Tahun Terbit: %d\n", daftarBuku[i].tahunTerbit);
    	printf("Penulis: %s\n", daftarBuku[i].penulis);
    	printf("Jumlah Halaman: %d\n", daftarBuku[i].halaman);
    	printf("Berat Buku: %lf\n", daftarBuku[i].berat);
    	printf("ISBN: %lld\n", daftarBuku[i].isbn);
    	printf("Rating: %f\n", daftarBuku[i].rating);
    	printf("Dipinjam? %s\n", (daftarBuku[i].dipinjam) ? "Iya" : "Tidak");
    	printf("-----------------------------\n");
    }
    printf("\nTekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}

void kembalikanBuku(Buku **daftarBuku, int *jumlahBuku, Riwayat **daftarRiwayat, int *jumlahRiwayat)
{
    int indexBuku = 0;
    char judul[128];

    printf("Masukkan nama buku yang ingin dikembalikan: ");
    scanf("%[^\n]", judul);
    getchar();

    if ((indexBuku = binarySearch(*daftarBuku, judul, 0, *jumlahBuku - 1)) == -1)
    {
        printf("Buku tidak ditemukan!\n");
        return;
    }


    if ((*daftarBuku)[indexBuku].dipinjam != 1)
    {
        printf("Buku tidak sedang dipinjam!\n");
        printf("Tekan tombol apa saja untuk kembali ke menu\n");
        getchar();
    }

    FILE *tmpFile = fopen("tmp", "w");

    for (int i = 0; i < *jumlahBuku; i++)
    {
        if (*jumlahBuku == indexBuku)
        {
            fprintf(tmpFile, "%s#%d#%s#%d#%lf#%lld#%f#0\n", (*daftarBuku)[i].judul,
                    (*daftarBuku)[i].tahunTerbit, (*daftarBuku)[i].penulis,
                    (*daftarBuku)[i].halaman, (*daftarBuku)[i].berat, (*daftarBuku)[i].isbn,
                    (*daftarBuku)[i].rating);
            continue;
        }
        fprintf(tmpFile, "%s#%d#%s#%d#%lf#%lld#%f#%i\n", (*daftarBuku)[i].judul,
                (*daftarBuku)[i].tahunTerbit, (*daftarBuku)[i].penulis,
                (*daftarBuku)[i].halaman, (*daftarBuku)[i].berat, (*daftarBuku)[i].isbn,
                (*daftarBuku)[i].rating, (*daftarBuku)[i].dipinjam);
    }

    fclose(tmpFile);
    tmpFile = fopen("tmp", "r");

    FILE *dbFile = fopen("db.txt", "w");
    char buffer[512];
    while (fscanf(tmpFile, "%[^\n]\n", buffer) != EOF)
    {
        fprintf(dbFile, "%s\n", buffer);
    }

    fclose(tmpFile);
    fclose(dbFile);
    remove("tmp");

    char peminjam[64];
    printf("Nama peminjam: ");
    scanf("%s", peminjam);
    getchar();

    time_t rawtime;
    struct tm *info;
    char date[80];

    time(&rawtime);

    info = localtime(&rawtime);

    strftime(date, 80, "%Y-%m-%d", info);

    FILE *histFile = fopen("history.txt", "a");
    fprintf(histFile, "%s#%s#%s#K\n", date, judul, peminjam);
    fclose(histFile);

    updateDatabase("db.txt", daftarBuku, jumlahBuku);
    updateRiwayat("history.txt", daftarRiwayat, jumlahRiwayat);

    printf("Buku telah dikembalikan!\n");
    printf("Tekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}

void pinjamBuku(Buku **daftarBuku, int *jumlahBuku, Riwayat **daftarRiwayat, int *jumlahRiwayat)
{
    int indexBuku = 0;
    char judul[128];

    printf("Masukkan nama buku yang ingin dipinjam: ");
    scanf("%[^\n]", judul);
    getchar();

    if ((indexBuku = binarySearch(*daftarBuku, judul, 0, *jumlahBuku - 1)) == -1)
    {
        printf("Buku tidak ditemukan!\n");
        return;
    }


    if ((*daftarBuku)[indexBuku].dipinjam != 0)
    {
        printf("Buku sedang dipinjam!\n");
        printf("Tekan tombol apa saja untuk kembali ke menu\n");
        getchar();
    }

    FILE *tmpFile = fopen("tmp", "w");

    for (int i = 0; i < *jumlahBuku; i++)
    {
        if (*jumlahBuku == indexBuku)
        {
            fprintf(tmpFile, "%s#%d#%s#%d#%lf#%lld#%f#1\n", (*daftarBuku)[i].judul,
                    (*daftarBuku)[i].tahunTerbit, (*daftarBuku)[i].penulis,
                    (*daftarBuku)[i].halaman, (*daftarBuku)[i].berat, (*daftarBuku)[i].isbn,
                    (*daftarBuku)[i].rating);
            continue;
        }
        fprintf(tmpFile, "%s#%d#%s#%d#%lf#%lld#%f#%i\n", (*daftarBuku)[i].judul,
                (*daftarBuku)[i].tahunTerbit, (*daftarBuku)[i].penulis,
                (*daftarBuku)[i].halaman, (*daftarBuku)[i].berat, (*daftarBuku)[i].isbn,
                (*daftarBuku)[i].rating, (*daftarBuku)[i].dipinjam);
    }

    fclose(tmpFile);
    tmpFile = fopen("tmp", "r");

    FILE *dbFile = fopen("db.txt", "w");
    char buffer[512];
    while (fscanf(tmpFile, "%[^\n]\n", buffer) != EOF)
    {
        fprintf(dbFile, "%s\n", buffer);
    }

    fclose(tmpFile);
    fclose(dbFile);
    remove("tmp");

    char peminjam[64];
    printf("Nama peminjam: ");
    scanf("%s", peminjam);
    getchar();

    time_t rawtime;
    struct tm *info;
    char date[80];

    time(&rawtime);

    info = localtime(&rawtime);

    strftime(date, 80, "%Y-%m-%d", info);

    FILE *histFile = fopen("history.txt", "a");
    fprintf(histFile, "%s#%s#%s#P\n", date, judul, peminjam);
    fclose(histFile);

    updateDatabase("db.txt", daftarBuku, jumlahBuku);
    updateRiwayat("history.txt", daftarRiwayat, jumlahRiwayat);

    printf("Buku telah dipinjam!\n");
    printf("Tekan tombol apa saja untuk kembali ke menu\n");
    getchar();
}